### PR TITLE
Construction Hook

### PR DESCRIPTION
**Pull request by joergkiesewetter**
adding a construction hook to the building kind interface

### CODE REVIEW

File: contracts/src/rules/BuildingRule.sol [hunk 1]
```diff
@@ -231,6 +231,16 @@ contract BuildingRule is Rule {
...
         }
+
+        // calls the preflight hook in the building implementation to add the possibility to make conditional builds 
+        // or consume inventory while building
+        BuildingKind buildingImplementation = BuildingKind(state.getImplementation(buildingKind));
+        // if no implementation set, then this is a no-op
+        if (address(buildingImplementation) == address(0)) {
+            return;
+        }
+        buildingImplementation.construct(game, buildingKind, mobileUnit, abi.encode(coords));
```
Comment (chrisfarms) [inline contracts/src/rules/BuildingRule.sol:242]: Some buildings don't have contracts ... but you still want them to get built ... so bailing earlier here will prevent those kinds of buildings ever getting contracted ... I think you want something like:

```suggestion
        if (address(buildingImplementation) != address(0)) {
            buildingImplementation.construct(game, buildingKind, mobileUnit, abi.encode(coords));
        }
```

**Review by 5p0rt5BEArD — APPROVED** [review]

**Comment by chrisfarms** [comment]
.... I'm just trying to work out a way this can get in without requiring breaking (or requiring updating of) existing all existing BuildingKinds 😅

**Comment by 5p0rt5BEArD** [comment]
> .... I'm just trying to work out a way this can get in without requiring breaking (or requiring updating of) all existing BuildingKinds 😅

Ah ok - I was assuming all was ok as this was already iterated on in the original PR

My approval was hasty

**Comment by joergkiesewetter** [comment]
sorry, didn't thought about this. would it make sense to make BuildingKind an abstract class?

**Pull request merged**